### PR TITLE
DOCSP-32418 Add OData tutorial

### PR DESCRIPTION
**Pull request by jordan-smith721**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-golang/blob/master/REVIEWING.md)

JIRA - https://jira.mongodb.org/browse/DOCSP-32418

### Staging Links

<li><a href=https://deploy-preview-286--mongodb-docs-csharp.netlify.app/fundamentals/odata>fundamentals/odata</a></li><li><a href=https://deploy-preview-286--mongodb-docs-csharp.netlify.app/fundamentals>fundamentals</a></li>

## Self-Review Checklist

- [x] Is this free of any warnings or errors in the RST?
- [x] Did you run a spell-check?
- [x] Did you run a grammar-check?
- [x] Are all the links working?
- [x] Are the [facets and meta keywords](https://wiki.corp.mongodb.com/display/DE/Docs+Taxonomy) accurate?

### CODE REVIEW

File: source/includes/fundamentals/code-examples/connection/OData.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    private readonly IQueryable<Restaurant> _restaurants;
+
+    public RestaurantsController(IMongoClient client)
+    {
+        var database = client.GetDatabase("sample_restaurants");
+        _restaurants = database.GetCollection<Restaurant>("restaurants")
+            .AsQueryable();
+    }
+
+    // Register Get endpoint and set max documents to 5
```
Comment (stephmarie17) [inline source/includes/fundamentals/code-examples/connection/OData.cs:21]: [s] I think these narrative comments can be updated to align with the [Code Comment Style Guide](https://docs.google.com/document/d/1Bnhh1OsGRUuXo_6yVUASGWiSX6sWuqrEFxPP6sLZDHM/edit?tab=t.0#heading=h.xv6glj2lnzjh)
```suggestion
    // Registers Get endpoint and sets max documents to 5
```

File: source/includes/fundamentals/code-examples/connection/OData.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+// start-configure
+using Microsoft.AspNetCore.OData;
+using Microsoft.OData.ModelBuilder;
+using MongoDB.Bson.Serialization.Conventions;
+using MongoDB.Driver;
+using ODataTest.Models;
+
+var builder = WebApplication.CreateBuilder(args);
+
+// Register a convention pack to convert fields to camel case
```
Comment (stephmarie17) [inline source/includes/fundamentals/code-examples/connection/OData.cs:39]: See above

File: source/includes/fundamentals/code-examples/connection/OData.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+var camelCaseConvention = new ConventionPack {
+        new CamelCaseElementNameConvention()
+    };
+ConventionRegistry.Register(
+    "CamelCase", camelCaseConvention, type => true);
+
+builder.Services.AddSingleton<IMongoClient>(
+    new MongoClient("Your connection URI"));
+
+// Register the Restaurants entity and set the Id field as the key
```
Comment (stephmarie17) [inline source/includes/fundamentals/code-examples/connection/OData.cs:49]: See above

File: source/fundamentals/odata.txt [hunk 4]
```diff
@@ -0,0 +1,203 @@
...
+      service and map your controller endpoints.
+
+      .. literalinclude:: /includes/fundamentals/code-examples/connection/OData.cs
+         :language: csharp
+         :start-after: // start-configure
+         :end-before: // end-configure
+
+      .. note::
+
+         Replace the ``<"Your connection URI">`` placeholder with your MongoDB connection string.
```
Comment (stephmarie17) [inline source/fundamentals/odata.txt:117]: [s] Update the code snippet to also use `< >` to match this instruction.

File: source/includes/fundamentals/code-examples/connection/OData.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+// Register a convention pack to convert fields to camel case
+var camelCaseConvention = new ConventionPack {
+        new CamelCaseElementNameConvention()
+    };
+ConventionRegistry.Register(
+    "CamelCase", camelCaseConvention, type => true);
+
+builder.Services.AddSingleton<IMongoClient>(
+    new MongoClient("Your connection URI"));
```
Comment (stephmarie17) [inline source/includes/fundamentals/code-examples/connection/OData.cs:47]: See above

File: source/fundamentals/odata.txt [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+      
+      This code performs the following actions:
+
+      - Instantiates a new ``MongoClient`` and passes it to the controller
```
Comment (stephmarie17) [inline source/fundamentals/odata.txt:121]: [n] Add punctuation to align with the rest of the list.
```suggestion
      - Instantiates a new ``MongoClient`` and passes it to the controller.
```

File: source/fundamentals/odata.txt [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+      - Defines the Entity Data Model (EDM) and registers ``Restaurants`` as an
+        entity set with the key, ``Id``.
```
Comment (stephmarie17) [inline source/fundamentals/odata.txt:123]: [n] I don't think this comma is necessary:
```suggestion
        entity set with the key ``Id``.
```

File: source/fundamentals/odata.txt [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+      .. code-block:: none
+
+         info: Microsoft.Hosting.Lifetime[14]
+               Now listening on: http://localhost:5183
+         info: Microsoft.Hosting.Lifetime[0]
+               Application started. Press Ctrl+C to shut down.
+         info: Microsoft.Hosting.Lifetime[0]
+               Hosting environment: Development
+         info: Microsoft.Hosting.Lifetime[0]
+               Content root path: /Users/jordan.smith/Documents/mongo/ODataTest/ODataTest
```
Comment (stephmarie17) [inline source/fundamentals/odata.txt:150]: [q] Should this be an [example name](https://www.mongodb.com/docs/meta/style-guide/style/names/) or do you not mind having your name here? 😄

**Review by stephmarie17 — APPROVED** [review]
LGTM with a few comments! Also not sure if it's just me, but in the preview there's a blank link in the ``Fundamentals`` TOC under ``Time Series Collections``. Not part of the diff but just noting in case that should be addressed.

File: source/fundamentals/odata.txt (hunk 4 above)
Comment (jordan-smith721) [reply source/fundamentals/odata.txt:117]: Good catch!

File: source/fundamentals/odata.txt (hunk 8 above)
Comment (jordan-smith721) [reply source/fundamentals/odata.txt]: Good call, updated to be a placeholder instead!

File: source/fundamentals/odata.txt [hunk 9]
```diff
@@ -0,0 +1,203 @@
...
+         :start-after: // start-configure
+         :end-before: // end-configure
+
+      .. note::
+
+         Replace the ``<"Your connection URI">`` placeholder with your MongoDB connection string.
+      
+      This code performs the following actions:
+
+      - Instantiates a new ``MongoClient`` and passes it to the controller.
```
Comment (sanych-sun) [inline source/fundamentals/odata.txt:121]: Instantiates a new ``MongoClient`` and register it as a singleton in the DI container.

File: source/fundamentals/odata.txt [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+             ...
+
+             ]
+         }
+
+Additional Information
+----------------------
+
+To learn more about ASP.NET Core OData, see the `Microsoft OData documentation
+<https://learn.microsoft.com/en-us/odata/webapi-8/overview>`__.
```
Comment (sanych-sun) [inline source/fundamentals/odata.txt:203]: It might makes sense to add https://www.odata.org/documentation/ as well.

Comment (sanych-sun) [inline source/fundamentals/odata.txt:203]: Also we might want to mention that current version does not support OData-Aggregation ($apply query option).

**Review by sanych-sun — CHANGES_REQUESTED** [review]

File: source/fundamentals/odata.txt (hunk 10 above)
Comment (jordan-smith721) [reply source/fundamentals/odata.txt]: Added a note to the section above where query operations are added

**Review by sanych-sun — APPROVED** [review]
LGTM

**Pull request merged**